rocsp/rocsp.go

```diff
@@ -270,3 +284,99 @@ func (c *Client) GetMetadata(ctx context.Context, serial string) (*Metadata, err
 	c.getLatency.With(prometheus.Labels{"result": "success", "method": "GetMetadata"}).Observe(time.Since(start).Seconds())
 	return &metadata, nil
 }
+
+// ScanResponsesResult represents a single OCSP response entry in redis.
+// `Serial` is the stringified serial number of the response. `Body` is the
+// DER bytes of the response. If this object represents an error, `Err` will
+// be non-nil and the other entries will have their zero values.
+type ScanResponsesResult struct {
+	Serial string
+	Body   []byte
+	Err    error
+}
+
+// ScanResponses scans Redis for all OCSP responses where the serial number matches the provided pattern.
+// It returns immediately and emits results and errors on `<-chan ScanResponsesResult`. It closes the
+// channel when it is done or hits an error.
+func (c *Client) ScanResponses(ctx context.Context, serialPattern string) <-chan ScanResponsesResult {
+	pattern := fmt.Sprintf("r{%s}", serialPattern)
+	results := make(chan ScanResponsesResult)
+	go func() {
+		defer close(results)
+		var cursor uint64
+		for {
+			var keys []string
+			var err error
+			keys, cursor, err = c.rdb.Scan(ctx, cursor, pattern, 10).Result()
+			if err != nil {
+				results <- ScanResponsesResult{Err: err}
+				return
+			}
+			if cursor == 0 {
+				return
+			}
+			for _, key := range keys {
+				serial, err := SerialFromResponseKey(key)
+				if err != nil {
+					results <- ScanResponsesResult{Err: err}
+					return
+				}
+				val, err := c.rdb.Get(ctx, key).Result()
```

Comment:
As per comment on line 372, should this be `.GetResponse(ctx, serial)` instead?

Reply:
As of right now, what `scan-responses` does is print the base64 of the responses, so parsing the response would be somewhat redundant - and would probably serve to make the scan CPU-bound. It is somewhat nice to have a fast tool that simply spits out the data, so we can compose it with other tools that (for instance) parallelize parsing and looking for specific things.